readthedocs/core/unresolver.py

```diff
@@ -1,99 +1,279 @@
+            return None
 
-        # Handle returning a response
-        if hasattr(project_slug, 'status_code'):
+        canonical_project = Project.objects.filter(slug=project_slug).first()
+        if not canonical_project:
             return None
 
-        request.host_project_slug = request.slug = project_slug
-        return self.unresolve_from_request(request, url)
+        project, version, filename = self._unresolve_path(
+            canonical_project=canonical_project,
+            path=parsed.path,
+        )
+
+        if add_index and filename and filename.endswith("/"):
+            filename += "index.html"
 
-    def unresolve_from_request(self, request, path):
+        return UnresolvedURL(
+            canonical_project=canonical_project,
+            project=project or canonical_project,
+            version=version,
+            filename=filename,
+            query=parsed.query,
+            fragment=parsed.fragment,
+            domain=domain_object,
+            external=external,
+        )
+
+    @staticmethod
+    def _normalize_filename(filename):
+        """Normalize filename to always start with ``/``."""
+        filename = filename or "/"
+        if not filename.startswith("/"):
+            filename = "/" + filename
+        return filename
+
+    def _match_multiversion_project(self, canonical_project, path):
         """
-        Unresolve using a request.
+        Try to match a multiversion project.
 
-        ``path`` can be a full URL, but the domain will be ignored,
-        since that information is already in the request object.
+        If the translation exists, we return a result even if the version doesn't,
+        so the translation is taken as the canonical project (useful for 404 pages).
+        """
+        match = self.multiversion_pattern.match(path)
+        if not match:
+            return None
+
+        language = match.group("language")
+        version_slug = match.group("version")
+        file = self._normalize_filename(match.group("file"))
+
+        if canonical_project.language == language:
+            project = canonical_project
+        else:
+            project = canonical_project.translations.filter(language=language).first()
 
-        None is returned if the request isn't valid.
+        if project:
+            version = project.versions.filter(slug=version_slug).first()
+            if version:
+                return project, version, file
+            return project, None, None
+
+        return None
+
+    def _match_subproject(self, canonical_project, path):
         """
-        parsed = urlparse(path)
-        path = parsed.path
-        project_slug = getattr(request, 'host_project_slug', None)
+        Try to match a subproject.
 
-        if not project_slug:
+        If the subproject exists, we try to resolve the rest of the path
+        with the subproject as the canonical project.
+        """
+        match = self.subproject_pattern.match(path)
```

Comment:
```suggestion
        #  TODO: Allow the canonical_project to override this url somehow..
        # if canonical_project.subproject_pattern:
        #     match = canonical_project.subproject_pattern.match(path)
        match = self.subproject_pattern.match(path)
```

Is this the longer term goal for customization?

Reply:
yeah, basically on each call to match, will be some extra logic to use the custom pattern from the project instead of the default one.